Apply timezone to Timestamp when using bulkcopy for batch insert

tkyc opened the pull request:
Correctly applies the timezone to Timestamps when inserted using batch insert with bulkcopy.

Fixes https://github.com/microsoft/mssql-jdbc/issues/2271

PriyadharshiniP commented:
Hi @tkyc 

I appreciate you making the changes.

I observed that the fix had the following problems:
1. Could you kindly have a look at the applyTimezoneTo method, which appears to use 
`Date date = new Date(**cal**.getTimeInMillis());` rather than `Date date = new Date(**ts.getTime()**); }?
If you have a time value other than today's time in your unit test, **BatchExecutionTest** for **_testValidTimezoneForTimestampBatchInsertWithBulkCopy_**, it will begin to break. The unit test would have seemed to be successful because the timestamp and calendar will often have the same current time value. However, this unit test would begin to fail if you had a different timestamp value.
For example, the unit test failed when I modified the timestamp milliseconds from `long ms = Timestamp.valueOf(todayString).getTime();} to `long ms = 1433338533461L;`. that leads to the failure.

![image](https://github.com/microsoft/mssql-jdbc/assets/7245402/fd7c7263-8412-4c78-8671-59c753d4d6dc)

2. In the event that the batch insert for bulk copy API falls back to the original batch insert, this fix does not seem to function. An additional conversion occurs when the fallback to no bulk copy API is used, leading to a double conversion for GMT.
To replicate this issue, Use the existing `BulkCopyResultSetCursorTest` for `testMultiplePreparedStatementAndResultSet ` unit test method alone. Connection string should have  "useBulkCopyForBatchInsert=true;sendTemporalDataTypesAsStringForBulkCopy=false;". 
Because bulk copy does not allow IDENTITY columns, using bulk copy would fail and result in inaccurate database values with fallback.
![image](https://github.com/microsoft/mssql-jdbc/assets/7245402/d7d7125a-be44-45a5-bb3f-0f6e8397fb34)

tkyc commented:
@PriyadharshiniP Thanks for the feedback. I'll do more investigating and I'll look into correcting the implementation.

> May I ask if the changes need to be done at SQLServerBulkCopy ?

That was my initial thought also, but when I looked over SQLServerBulkCopy it doesn't look like I'm able to apply the timezone in the TDS request for bulkcopy.

For example, I sort of expected code like this in SQLServerBulkCopy, which is part of the normal and regular batch insert code path. The code below takes into consideration the the timezone by sending the timezone info in the TDS request. So the timezone is applied during insertion by the server:

```
                                    tdsWriter.writeEncryptedRPCDateTime2(name,
                                            timestampNormalizedCalendar(calendar, javaType, conn.baseYear()),
                                            subSecondNanos, (valueLength), isOutParam, statement);
```

But I'm guessing because we're doing a bulkcopy, it has to send the timestamp value as is. And so, either the user or driver needs to apply the timezone conversion beforehand.

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -54,6 +60,8 @@ public class BatchExecutionTest extends AbstractTest {
     private static String ctstable3;
     private static String ctstable4;
     private static String ctstable3Procedure1;
+    private static String timestampTable = AbstractSQLGenerator
```
this table should be added to dropTable() for cleanup

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
+                    + "') and OBJECTPROPERTY(id, N'IsUserTable') = 1) DROP TABLE " + timestampTable;
+            stmt.execute(dropSql);
+
+            String createSql = "CREATE TABLE" + timestampTable + " (c1 DATETIME2(3))";
```
this table should be added to dropTable() or add finally to cleanup

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
+        try (Connection con = DriverManager.getConnection(connectionString); Statement stmt = con.createStatement();
+                PreparedStatement pstmt = con.prepareStatement("INSERT INTO " + timestampTable + " VALUES(?)")) {
+
+            String dropSql = "IF EXISTS (SELECT * FROM dbo.sysobjects WHERE id = object_id(N'"
```
dropTableIfExists

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -13,11 +13,17 @@
 import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.TimeZone;
```
pls cleanup unused

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
+            rs.next();
+            ts1 = rs.getTimestamp(1);
+
+            assertEquals(ts0, ts1);
```
maybe add a bit more debug info if it fails? I think assertEquals will just tell u false it doesn't equal but not the values

tkyc replied:
`assertEquals` will output an expected vs. actual message, so no need for an actual failure message.

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -99,6 +107,55 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
+        }
+
+        try (Connection con = DriverManager.getConnection(connectionString); Statement stmt = con.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT * FROM " + timestampTable);
```
ResultSet is leaked

lilgreenbird commented on src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java:
```diff
@@ -2938,6 +2947,44 @@ private boolean checkSQLLength(int length) {
         return true;
     }
 
+    private String getDateFormatPattern(int nanos) {
```
add comments on what this does

lilgreenbird commented on src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java:
```diff
@@ -2938,6 +2947,44 @@ private boolean checkSQLLength(int length) {
+        return pattern.toString();
+    }
+
+    private java.sql.Timestamp applyTimezoneTo(java.sql.Timestamp ts, Calendar cal) {
```
add comment for method

PriyadharshiniP commented:
> 1. **BatchExecutionTest**

Thanks for fixing the issue.
1.  Additionally, it appears that not all dates—particularly those with daylight savings time—cannot be converted successfully using the applyTimezoneTo method . In your BatchExecutionTest unit test case , if you change the ms to ` long ms = 1696127400000L;` that equates to :

```
GMT: Sunday, October 1, 2023 2:30:00 AM
My local time zone: Sunday, October 1, 2023 1:30:00 PM [GMT+11:00]
```

This time "GMT: Sunday, October 1, 2023 2:30:00 AM" doesn't exist in my local time zone. The non bulk version correctly inserts the date as `2023-10-01 02:30:00.000` but the bulk version incorrectly inserts as `2023-10-01 03:30:00.000` and the BatchExecutionTest  will fail if we use  ` ts0 = rs.getTimestamp(1, gmtCal);. and ts1 = rs.getTimestamp(1, gmtCal);`
I think this is because of using Timestamp.valueOf or any timestamp related methods that will use only local time for conversions.

![image](https://github.com/microsoft/mssql-jdbc/assets/7245402/3c97ba8f-93f2-4a04-8beb-a807e23491c3)

2. As the standard batch insert without bulk copy seems to work in all cases, I started to look re-use the tdsWriter.writeEncryptedRPCDateTime2 in SQLServerBulkCopy itself, however while attempting to fix it, I encounter certain problems.
To store the parameters needed for the writeEncryptedRPCDateTime2 method, I made a new class. Next, update the SQLServerBulkBatchInsertRecord class with those values. Fill in the datetime2 field in SQLServerBulkCopy in accordance with the screenshots below :
![image](https://github.com/microsoft/mssql-jdbc/assets/7245402/0394bf94-a82f-4cc0-8971-250ea38dfc1d)
![image](https://github.com/microsoft/mssql-jdbc/assets/7245402/5ec7c661-e4a4-490e-a1b5-6d6ba93157fb)
However, the following error appears:  `java.sql.BatchUpdateException: While reading current row from host, a premature end-of-message was encountered--an incoming data stream was interrupted when the server expected to see more data. The host program may have terminated. Ensure that you are using a supported client application programming interface (API).` 
Kindly inform me if you can assist with this.

Could you see if we can accomplish something along these lines?

tkyc commented:
@PriyadharshiniP Thanks for collaborating.

> As the standard batch insert without bulk copy seems to work in all cases, I started to look re-use the tdsWriter.writeEncryptedRPCDateTime2 in SQLServerBulkCopy

Yeah, I struggled with getting that working as well when I initially looked at the code. In `SQLServerBulkCopy`, it seems the only thing written to the TDS request is column metadata info and data type info and it doesn't look like any input values are written eg. the timestamp in our case for batch insert with bulkcopy. When I traced the batch insert bulkcopy call path:

```
writeToServer -> sendBulkLoadBCP -> ... -> doInsertBulk -> sendBulkCopyCommand
```

In the above call path, I expected to end up somewhere where the timestamp value is written to the TDS request (something/somewhere where the logic `writeEncryptedRPCDateTime2 ` can be used) but I don't. I'll do some more debugging, maybe I missed some things initially.

> Additionally, it appears that not all dates—particularly those with daylight savings time—cannot be converted successfully using the applyTimezoneTo method

Thanks for the help on testing. I'll see what I can do here...

tkyc commented:
@PriyadharshiniP What's your exact local time zone? I've switched my PC's time zone setting to various different time zones that doesn't have day light savings and wasn't able to repro the issue. I double checked the local time zone used in my JDK and it matches my PC time zone.

PriyadharshiniP commented:
My local timezone is Australia/Sydney

On Wed, 17 Jan 2024, 8:11 am Terry Chow, ***@***.***> wrote:

> @PriyadharshiniP <https://github.com/PriyadharshiniP> What's your exact
> local time zone? I've switched my PC's time zone setting to various
> different time zones that doesn't have day light savings and wasn't able to
> repro the issue. I double checked the local time zone used in my JDK and it
> matches my PC time zone.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/mssql-jdbc/pull/2291#issuecomment-1894518846>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXI4WXKWWK5KQJQM6BSDXLYO3UHDAVCNFSM6AAAAABBVM2HUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJUGUYTQOBUGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

PriyadharshiniP commented:
> @PriyadharshiniP What's your exact local time zone? I've switched my PC's time zone setting to various different time zones that doesn't have day light savings and wasn't able to repro the issue with `1696127400000L`. I double checked the local time zone used in my JDK and it matches the PC time zone set when I was experimenting.
> 
> Maybe I need to do something like below explicitly to repro your issue?
> 
> ```
> TimeZone tz = TimeZone.getTimeZone("your-time-zone");
> TimeZone.setDefault(tz);
> ```

I guess you could do something like this in your unit test  :`System.setProperty("user.timezone","Australia/Sydney");`

tkyc commented:
@PriyadharshiniP I've just pushed a new fix implementation. I did some testing of it and it seem to have addressed all your concerns so far. Let me know if there's any additional problems. Thanks again for the collab.

Jeffery-Wasty commented on src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java:
```diff
@@ -2480,10 +2480,13 @@ else if (4 >= bulkScale)
                                 // Send only the date part
-                                tdsWriter.writeDate(timeStampValue.substring(0, timeStampValue.lastIndexOf(' ')));
+                                tdsWriter.writeDate(ts.getTime(), cal);
+                                //tdsWriter.writeRPCDateTime2(name,
+                                //        timestampNormalizedCalendar(calendar, javaType, conn.baseYear()),
+                                //        subSecondNanos, TDS.MAX_FRACTIONAL_SECONDS_SCALE, isOutParam);
```
Remove commented out code

Jeffery-Wasty commented on src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java:
```diff
@@ -3836,6 +3850,22 @@ void writeTime(java.sql.Timestamp value, int scale) throws SQLServerException {
 
+    void writeTime(java.sql.Timestamp value, int scale, Calendar cal) throws SQLServerException {
+        GregorianCalendar calendar = initializeCalender(TimeZone.getDefault());
+        long utcMillis; // Value to which the calendar is to be set (in milliseconds 1/1/1970 00:00:00 GMT)
+        int subSecondNanos;
+        utcMillis = value.getTime();
+        subSecondNanos = value.getNanos();
```
Declarations and assignments can be combined.
```suggestion
        long utcMillis = value.getTime(); // Value to which the calendar is to be set (in milliseconds 1/1/1970 00:00:00 GMT)
        int subSecondNanos = value.getNanos();
```

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -54,6 +60,10 @@ public class BatchExecutionTest extends AbstractTest {
+    private static String timestampTable1 = AbstractSQLGenerator
+            .escapeIdentifier(RandomUtil.getIdentifier("timestamptable1"));
+    private static String timestampTable2 = AbstractSQLGenerator
+            .escapeIdentifier(RandomUtil.getIdentifier("timestampTableBatchInsert2"));
```
stupid nitpick, usually we use same identifier as the name.

lilgreenbird commented on src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BatchExecutionTest.java:
```diff
@@ -99,6 +109,186 @@ public void testBatchUpdateCountTrueOnFirstPstmtSpPrepare() throws Exception {
+
+        // Compare Timestamp values inserted, should be the same
+        try (Connection con = DriverManager.getConnection(connectionString); Statement stmt = con.createStatement()) {
+            try (ResultSet rs = stmt.executeQuery("SELECT * FROM " + timestampTable1)) {
```
why not combine to same try?

PriyadharshiniP commented:
> @PriyadharshiniP I've just pushed a new fix implementation. I did some testing of it and it seem to have addressed all your concerns so far. Let me know if there's any additional problems. Thanks again for the collab.

Hi @tkyc 
Thanks for fixing it so quickly . 
I checked all dates this year in my local time zone, and it seems that there will still be problems with the bulk copy inserting the date in my local environment around spring forward dates. The time stamp millisecond values are 1712415600000L and 1743935400000L, to be more precise.
Noted the use of `java.sql.Timestamp.valueOf(colValue.toString())` in SQLServerBulkCopy, this seems to be the cause of this issue as it seems to alter the timestamp milliseconds value. Using the `(Timestamp)colValue` seems to work for all dates. Could you please check and fix this as well?

![image](https://github.com/microsoft/mssql-jdbc/assets/7245402/cd85511c-9989-4d27-a330-a1716cced9be)
![image](https://github.com/microsoft/mssql-jdbc/assets/7245402/8ad4331e-7bea-4860-be9f-4c832cacc607)

tkyc commented:
@PriyadharshiniP Oops, my mistake. Calling valueOf will cause a conversion. Fixed.

PriyadharshiniP commented:
> @PriyadharshiniP Oops, my mistake. Calling valueOf will cause a conversion. Fixed.

Thanks @tkyc :
It seems to be fine for my tests.
Please let me  know when the release would be available ?

Regards

tkyc commented:
@PriyadharshiniP this will make it in the release that's on the 31st of this month.
